vsphere_plugin_common/__init__.py

```diff
@@ -596,139 +765,527 @@ def get_server_list(self):
+                ctx.logger.warn(
+                    'Host {host} does not have resource pool {rp}.'.format(
+                        host=host.name,
+                        rp=resource_pool,
+                    )
+                )
+                continue
+
+            host_nets = set([
+                (
+                    network['name'],
+                    network['switch_distributed'],
+                )
+                for network in self.get_host_networks(host)
+            ])
+            vm_nets = set([
+                (
+                    network['name'],
+                    network['switch_distributed'],
+                )
+                for network in vm_networks
+            ])
+
+            nets_not_on_host = vm_nets.difference(host_nets)
+
+            if nets_not_on_host:
+                message = 'Host {host} does not have all required networks. '
+
+                missing_standard_nets = ', '.join([
+                    net[0] for net in nets_not_on_host
+                    if not net[1]
+                ])
+                missing_distributed_nets = ', '.join([
+                    net[0] for net in nets_not_on_host
+                    if net[1]
+                ])
+
+                if missing_standard_nets:
+                    message += 'Missing standard networks: {nets}. '
+
+                if missing_distributed_nets:
+                    message += 'Missing distributed networks: {dnets}. '
+
+                ctx.logger.warn(
+                    message.format(
+                        host=host.name,
+                        nets=missing_standard_nets,
+                        dnets=missing_distributed_nets,
+                    )
+                )
+                continue
+
+            ctx.logger.debug(
+                'Host {host} is a candidate for deployment.'.format(
+                    host=host.name,
+                )
+            )
+            candidate_hosts.append(host)
+
+        # Sort hosts based on the best processor ratio after deployment
+        candidate_hosts = [
+            (
+                host,
+                self.host_cpu_thread_usage_ratio(host, vm_cpus)
+            ) for host in candidate_hosts
+        ]
+        if candidate_hosts:
+            ctx.logger.debug(
+                'Host CPU ratios: {ratios}'.format(
+                    ratios=', '.join([
+                        '{hostname}: {ratio}'.format(
+                            hostname=c[0].name,
+                            ratio=c[1]
+                        ) for c in candidate_hosts
+                    ])
+                )
+            )
+        candidate_hosts.sort(
+            reverse=True,
+            key=lambda host_rating: host_rating[1]
+        )
+        candidate_hosts = [
+            host[0] for host in candidate_hosts
+        ]
+
+        if candidate_hosts:
+            return candidate_hosts
+        else:
+            message = (
```

Comment:
I find them less ugly than the alternative for splitting that over multiple lines.

It'd be possible to format it in one if that's preferred?